File: day02.py
import math
import re
from collections import defaultdict
from pathlib import Path
from pprint import pprint

Input = dict[int, list[dict[str, int]]]

INPUT_FILE_PATH = Path("input.txt")

COLOR_MAXES = {
    "red": 12,
    "green": 13,
    "blue": 14,
}


def read_input() -> Input:
    data = {}
    with open(INPUT_FILE_PATH, "r", encoding="utf-8") as f:
        for line in f:
            line = line.strip()
            if line == "":
                continue
            match = re.match(r"Game (\d+):", line)
            game_num = int(match.group(1))
            pulls_raw = line[match.end():].split(";")
            pulls = []
            for pull in pulls_raw:
                counts = defaultdict(int)
                for color in COLOR_MAXES.keys():
                    if (match := re.search(f"(\\d+) {color}", pull)) is not None:
                        counts[color] = int(match.group(1))
                pulls.append(counts)
            data[game_num] = pulls
    return data


def solve1(input_: Input) -> int:
    good_game_nums = set()
    for game_num, pulls in input_.items():
        if all(
            all(pull[color] <= count_max for color, count_max in COLOR_MAXES.items())
            for pull in pulls
        ):
            good_game_nums.add(game_num)
    return sum(good_game_nums)


def solve2(input_: Input) -> int:
    total_power = 0
    for game_num, pulls in input_.items():
        biggest_seen = defaultdict(int)
        for pull in pulls:
            for color, count in pull.items():
                best_prev = biggest_seen[color]
                biggest_seen[color] = max(best_prev, count)
        power = math.prod((biggest_seen[color] for color in COLOR_MAXES))
        total_power += power
    return total_power


def main():
    input_ = read_input()
    answer = solve1(input_)
    pprint(answer)
    answer = solve2(input_)
    pprint(answer)


if __name__ == "__main__":
    main()

File: javascript/day02.ts
import * as fs from "fs";
import { stringify } from "querystring";

type Pair<T, U> = [T, U];
type Input = Map<number, Array<Pair<string, number>>>;

const INPUT_FILE_PATH = "input.txt";

const COLOR_MAXES = {
    red: 12,
    green: 13,
    blue: 14,
}
const COLORS = Object.keys(COLOR_MAXES);


function readInput(): Input {
    const fileString = fs.readFileSync(INPUT_FILE_PATH, "utf-8");
    const data: Input = new Map();
    for (let line of fileString.split("\n")) {
        line = line.trim();
        if (!line) {
            continue;
        }
        const regexGameNum = /Game (\d+):/g;
        const gameNumMatch = regexGameNum.exec(line);
        if (gameNumMatch == null) {
            continue;
        }
        const gameNum = parseInt(gameNumMatch[1]);
        const pullsRaw = line.substring(regexGameNum.lastIndex);
        const pulls = Array.from(pullsRaw.matchAll(new RegExp(`(\\d+)\\s+(${COLORS.join("|")})`, "g")))
            .map(match => [match[2], parseInt(match[1])] as Pair<string, number>);
        data.set(gameNum, pulls);
    }
    return data;
}


function solve1(input: Input): number {
    return Array.from(input.entries())
        .filter(([_, pulls]) => pulls.every(
            ([color, count]) => (count <= COLOR_MAXES[color])
        )).map(([gameNum, _]) => gameNum)
        .reduce((total, current) => total + current);
}


function solve2(input: Input): number {
    let totalPower = 0;
    for (const pulls of input.values()) {
        const biggestSeen = new Map<string, number>();
        for (const [color, count] of pulls) {
            const best_prev = biggestSeen.get(color) || 0;
            biggestSeen.set(color, Math.max(best_prev, count));
        }
        const power = COLORS.map(color => biggestSeen.get(color) || 0)
            .reduce((prod, num) => prod * num);
        totalPower += power;
    }
    return totalPower;
}


function main() {
    const input = readInput();
    const answer1 = solve1(input);
    console.log(answer1);
    const answer2 = solve2(input);
    console.log(answer2);
}


main();
